### PR TITLE
Initial implementation of documentation blacklist

### PR DESCRIPTION
**Pull request by vignesh-sankaran**
Initial work on #626.

### CODE REVIEW

File: src/krate.rs [hunk 1]
```diff
@@ -1032,6 +1033,10 @@ pub fn show(req: &mut Request) -> CargoResult<Response> {
     )?;
     let max_version = krate.max_version(&conn)?;
 
+    // Cheaper than cloning the value to the function
+    let doc_url = mem::replace(&mut krate.documentation, None);
```
Comment (sgrif) [inline src/krate.rs]: `let doc_url = krate.documentation.take()`?

**Comment by vignesh-sankaran** [comment]
@sgrif I talked to @carols10cents about this on IRC, I'll let them provide their opinion here but a problem that was raised with an upload-time approach is if the blacklisted URL isn't in the crate database entry, then a crate author asks why their documentation URL isn't showing up, it won't be immediately obvious.

**Comment by carols10cents** [comment]
Yeah I want the URL in the database for my own debugging purposes. What if, instead of calling `documentation_blacklist` from `fn show`, we call it in the `Crate` `encodable` method so that anywhere we encode a crate for returning via JSON, we've blacklisted documentation URLs?

**Comment by vignesh-sankaran** [comment]
Implemented the changes as suggested @carols10cents, was there anything else that needed to be discussed?

File: src/krate.rs [hunk 2]
```diff
@@ -545,6 +548,16 @@ impl Crate {
         }
     }
 
+    fn documentation_blacklist(url: Option<String>) -> Option<String> {
```
Comment (sgrif) [inline src/krate.rs]: Can we give this a more descriptive name? How about `remove_blacklisted_documentation_urls`

File: src/krate.rs [hunk 3]
```diff
@@ -1052,6 +1065,8 @@ pub fn show(req: &mut Request) -> CargoResult<Response> {
     )
 }
 
+
```
Comment (sgrif) [inline src/krate.rs]: Can you delete these extra blank lines?

File: src/krate.rs [hunk 4]
```diff
@@ -505,7 +505,7 @@ impl Crate {
             downloads,
             description,
             homepage,
-            documentation,
+            mut documentation,
```
Comment (sgrif) [inline src/krate.rs]: You can remove this `mut` and just do `let documentation =` below. We're never actually mutating anything.

**Comment by vignesh-sankaran** [comment]
Cleaned up the history here as well @sgrif, learned a lot about git :)

Regarding the line `let documentation = Crate::remove_blacklisted_documentation_urls(documentation);`,  is the `documentation` value being passed into the function a copy of the existing variable, or a reference?

**Comment by carols10cents** [comment]
> Regarding the line let documentation = Crate::remove_blacklisted_documentation_urls(documentation);, is the documentation value being passed into the function a copy of the existing variable, or a reference?

It's transferring ownership of the original documentation value to the `Crate::remove_blacklisted_documentation_urls` function; technically it does a copy but that might get optimized away. If it was a reference, the signature of `Crate::remove_blacklisted_documentation_urls` would have `&` in front of its parameters' types, and you'd call it with `&documentation`. The [ownership chapter of the book](https://doc.rust-lang.org/nightly/book/second-edition/ch04-00-understanding-ownership.html) goes into this in more detail :)

It looks like you might be done with this, so I'm going to take a look :)

File: src/krate.rs [hunk 5]
```diff
@@ -545,6 +547,16 @@ impl Crate {
         }
     }
 
+    fn remove_blacklisted_documentation_urls(url: Option<String>) -> Option<String> {
+        let blacklisted_url = "rust-ci.org";
```
Comment (carols10cents) [inline src/krate.rs]: This is a great start!! Thank you!! Two improvements I'd like to see before we merge this in, however:

- I think a `contains` is a bit overzealous. While I'm definitely making up a silly case, this implementation would block someone who had their docs published at http://rust-ci.organists.com, and that would be confusing and unexpected both for crate users and for me trying to figure out what was going on. I think we should use [the url crate's `parse` function and `host_str` method](https://docs.rs/url/1.5.1/url/) to parse the documentation URL (if parsing fails, return `None`), extract the host as a string slice, and then compare against the blacklist for exact match.
- It'd be super nice if instead of one hardcoded URL, we had a hardcoded list here (that would contain one item) so that adding another URL to this blacklist would be adding that to the list, not having to change the code. The blacklist could be a constant at the top of the file so that we're not allocating the list every time we encode a crate, something like:

```
const DOCUMENTATION_BLACKLIST: [&'static str; 1] = [
    "rust-ci.org",
];
```

Then in the code, since with the first change we'll have just the host string, we actually *could* use `contains`, since then we'd be asking if the list contains the host exactly :)

Do these changes make sense? Please let me know if you have any questions!

**Comment by vignesh-sankaran** [comment]
Thank you to SneakySnake on IRC for helping me with the latest commit, had some difficulties with handling this many cases before running the blacklisting check :)

File: src/krate.rs [hunk 6]
```diff
@@ -547,13 +549,28 @@ impl Crate {
         }
     }
 
+    // Return None if the documentation URL host matches a blacklisted host
     fn remove_blacklisted_documentation_urls(url: Option<String>) -> Option<String> {
-        let blacklisted_url = "rust-ci.org";
+        let url = match url {
+            Some(url) => url,
+            None => return None,
+        };
```
Comment (carols10cents) [inline src/krate.rs]: This is such a perfect case for when `?` is supported on `Option`! (nothing to change here, I'm just noting that I wish there was a prettier way to do this but there isn't so this is fine :))

File: src/krate.rs [hunk 7]
```diff
@@ -547,13 +549,28 @@ impl Crate {
...
+
+        // Handles the case if the documentation URL does not parse successfully
+        let parsed_url = match Url::parse(&url) {
+            Ok(parsed_url) => parsed_url,
+            Err(_) => return None,
+        };
```
Comment (carols10cents) [inline src/krate.rs]: I've been trying to see if we can use methods like `map` on the `Result`s and `Option`s in here, but it's just complex enough that I don't think the end result is any clearer than what you have here. I see what you mean about handling lots of cases!! (again nothing to do here I'm just talking)

File: src/krate.rs [hunk 8]
```diff
@@ -39,6 +39,8 @@ use util::{RequestUtils, CargoResult, internal, ChainError, human};
 use version::{EncodableVersion, NewVersion};
 use {Model, User, Keyword, Version, Category, Badge, Replica};
 
+const DOCUMENTATION_BLACKLIST: [&'static str; 1] = ["rust-ci.org"];
```
Comment (carols10cents) [inline src/krate.rs]: What do you think about adding a doc comment for this constant explaining why a host might be in this list?

File: src/krate.rs [hunk 9]
```diff
@@ -547,13 +549,28 @@ impl Crate {
         }
     }
 
+    // Return None if the documentation URL host matches a blacklisted host
```
Comment (carols10cents) [inline src/krate.rs]: Should this be a doc comment with 3 slashes instead?

File: src/krate.rs (hunk 6 above)
Comment (vignesh-sankaran) [inline src/krate.rs]: Hmm is there an RFC for this somewhere?

File: src/krate.rs (hunk 8 above)
Comment (vignesh-sankaran) [inline src/krate.rs]: Will do :)

File: src/krate.rs (hunk 9 above)
Comment (vignesh-sankaran) [inline src/krate.rs]: I can do that, but will documentation comments appear on private functions?

File: src/krate.rs (hunk 7 above)
Comment (vignesh-sankaran) [inline src/krate.rs]: I did try something fancier involving `if let`, but it was heading to an arrow anti-pattern and you can't have an `if let` statement with the second variable depending on the result of the first. This solution is a lot simpler :).

**Comment by vignesh-sankaran** [comment]
@carols10cents I agree with covering the edge cases, but I felt unit tests shouldn't be in the `tests/` directory since they are meant for integration tests. That, and I'd have to make `remove_blacklisted_documentation_urls()` a public function, which I don't see a need for outside the realm of testing. So I created a new tests module in `src/krate.rs` and put the unit tests there instead. Should we have a long term view of putting unit tests e.g. tests specifically for `Crate::valid_feature_name` into associated modules instead, rather than have them in the `tests/` directory?

File: src/krate.rs (hunk 6 above)
Comment (carols10cents) [inline src/krate.rs]: Yup! https://github.com/rust-lang/rfcs/blob/9c14315070678e71c01f82a41d50b2489747b62c/text/1859-try-trait.md

**Comment by carols10cents** [comment]
> Should we have a long term view of putting unit tests e.g. tests specifically for Crate::valid_feature_name into associated modules instead, rather than have them in the tests/ directory?

Yup, that sounds good to me! There are a few unit tests in a few files, like [category.rs](https://github.com/rust-lang/crates.io/blob/master/src/category.rs#L286). I would like if there were more :)

**Comment by carols10cents** [comment]
Looks good!!! Thank you!!!!! 🍂 📗 💯

**Comment by carols10cents** [comment]
@vignesh-sankaran are you on twitter?

**Comment by vignesh-sankaran** [comment]
Yay, thanks @carols10cents :). Yes I am on Twitter, my handle is @rustbackend.
